### ISSUE TITLE
[css-display] Why is display listed as not animatable instead of animation type: discrete?

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#6429]
https://www.w3.org/TR/css-display-3/#the-display-properties

I thought we were going for discrete for all properties that can't be intelligently interpolated, and that "not animatable" was reserved primarily for animation properties and the like?

Especially about `display`, there are plenty of use cases where authors want to finish an animation with `display: none`, and this prevents them from doing so.

**Comment by flackr** [comment]
`display` does affect animations in the same way as animation properties. Specifically from https://www.w3.org/TR/css-animations-1/#animations

> Setting the display property to none will terminate any running animation applied to the element and its descendants. If an element has a display of none, updating display to a value other than none will start all animations applied to the element by the animation-name property, as well as all animations applied to descendants with display other than none.

This means that if a CSS animation animated to `display: none` it would immediately cancel the animation, losing the `display: none` which would restart the animation.

**Comment by LeaVerou** [comment]
Ah, bummer. "Not animatable" makes total sense then.

**Comment by Loirooriol** [comment]
Just for reference, at some point `display` was animatable on Firefox, but that could freeze the browser. https://bugzil.la/1264396

**Comment by flackr** [comment]
I wonder if we could support only specific values of display, e.g. anything but `none` on css animations, and possibly anything on web animations since their lifetime is not controlled by the display property. Then you could create an animation which maintains `display: block` until it finishes or use a web animation.

**Comment by tabatkins** [comment]
Having only some values animateable would be a new concept, but yes, that would work and probably be useful.

**Comment by flackr** [comment]
Reopening this to consider allowing animations of display to prevent the element from immediately computing to `display: none`.

In particular, my straw-man proposal is the following:
1. Animation keyframes may specify display values which are not none. `display: none` is dropped from keyframes.
2. Display is interpolated like visibility, i.e. it will prefer the non-none value.

With these two points, we shouldn't have any circularity. An animation cannot itself produce `display: none` since it can't specify it (rule 1). Transitions to `display: none` would also work since the `display: none` wouldn't apply until after the transition finished (rule 2).

With this, developers could do things like:
```css
.hide {
  transition: opacity 200ms, display 200ms;
  display: none;
  opacity: 0;
}
```

Or:
```css
@keyframes slideaway {
  to { transform: translateY(40px); opacity: 0; display: block; }
}

.hide {
  animation: slideaway 200ms;
  display: none;
}
```

**Comment by chrishtr** [comment]
> With this, developers could do things like

Just to add on: this should also work for an "entry" animation from `display:none` to `display:block` with the same proposed change.

**Comment by Loirooriol** [comment]
It's worth noting that the outer display type can already be discretely animated (indirectly, and with side-effects), e.g. by setting `display: inline-grid` outside of animations, and then animating `position`/`float` to trigger a blockification into `display: grid`. See  https://github.com/w3c/csswg-drafts/issues/6846#issuecomment-983107012

**Comment by jakearchibald** [comment]
Sorry for the long post here, but I think there's some interactivity issues worth considering here.

Let's say the JavaScript developer on a project creates a dialog/form, and when the form is submitted, the form is hidden via `display: none`, and the result of the form is used somehow in the app.

Then, at some later point, the CSS developer adds a transition which extends the interactive life of the element beyond what the JavaScript developer expected (the JavaScript developer and CSS developer may be the same person, but their mindset may be different when working with styles vs script).

This can be done already today, by overriding the default `display: none` on `<dialog>` elements (https://jsbin.com/xumitir/edit?html,css,js,output), but animating `display: none` creates an even more attractive, less-hacky way to do this.

By doing this, the CSS developer has created a significant change to the behaviour of the app. The user can now provide more than one submission to the form, since its lifetime in the app has been extended. This user action might be intentional (clicking the 'wrong' thing, then quickly clicking the right thing) or unintentional (double clicking).

This wasn't a possibility when the JavaScript developer created the form, so it's possible that these new interactions will be badly handled. The solution is to cater for interactions with the form beyond its previous lifecycle, or to prevent further interactions with the form using `inert`.

I think this is worth thinking about for two reasons:

- It's difficult to catch bugs resulting from this in development. My experience is developers rarely correctly cater for the extended lifetime of interactive elements during transitions-away.
- The issue is introduced by CSS, and the only way to fix it is in JS (`pointer-events: none` is an incomplete fix).

Potential solutions:

## Education

We tell developers about this problem, and how to solve it in JavaScript. However, I have low confidence that we'd reach everyone, and even when it comes to folks we'd reach, would they remember to avoid these bugs?

## Allow `inert` to be set in CSS

```css
.hide {
  transition: opacity 200ms, display 200ms;
  display: none;
  opacity: 0;
  interactivity: inert;
}
```

The current HTML `inert` attribute could be explained by a UA style:

```css
[inert] {
  interactivity: inert;
}
```

I know there was opposition to this in the past, since it relates to behaviour rather than style, but it feels like `display: none` and particularly transitions on `display: none` already have a significant impact on behaviour. Showing an element changes the behaviour of the page, and if that's done in CSS, it seems reasonable to be able to express modality via CSS.

Open question: Should it be possible for a child of an inert element to become un-inert? This is possible with `pointer-events`.

This means that the interactivity issue created by CSS can be solved with CSS. But, it requires the developer to remember to set it, so:

## Make elements transitioning to `display:none` inert by default


```css
.hide {
  transition: opacity 200ms, display 200ms;
  display: none;
  opacity: 0;
}
```

During the transition above, `interactivity` could behave as `inert`. This would avoid the situation described at the start of this post, where the CSS developer introduced a behaviour the JS developer hadn't catered for.

If the developer wants the element to be interactive during the transition, they can override this default:

```css
.hide {
  transition: opacity 200ms, display 200ms;
  display: none;
  opacity: 0;
  interactivity: interactive;
}
```

**Comment by yisibl** [comment]
Chrome 111.0.5545.0 is already supported and requires a Flag:
```
/Applications/Google\ Chrome\ Canary.app/Contents/MacOS/Google\ Chrome\ Canary --enable-blink-features=CSSDisplayAnimation
```

Demo: https://codepen.io/yisi/pen/RwBzqGE

https://user-images.githubusercontent.com/2784308/218027861-5774bf58-f95e-4723-912d-5d59fc8c63fb.mp4

**Comment by fantasai** [comment]
@flackr  Edited in [your proposal[(https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1318933547), see https://github.com/w3c/csswg-drafts/commit/b255d549bbb7a9df1975224d2a5b2c2cbc1ac068. One thing I'm not clear on is what happens to `display` declarations inside an `@keyframes`. It's clear that you intended them to be ignored, but are they dropped at parse time or do they appear in the OM?

**Comment by flackr** [comment]
`display` declarations in `@keyframes` would be used as is except for `display: none`. The current proposal is that `none` values would be replaced by `revert` per @emilio's suggestion in https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1332439874 . We haven't prototyped this yet - see [crbug.com/1411474](https://crbug.com/1411474). Alternately we could drop them from the processed keyframes but I'm not sure if that would raise other issues.

**Comment by Loirooriol** [comment]
Would `revert-layer` make more sense?

**Comment by flackr** [comment]
So I was worried about the circularity of an animation animating to a state that stops the animation which in turn restarts it but we might already have this with `content` being animatable on pseudo elements with something like this:

```css
@keyframes no-content {
  from { content: 'anim-start'; }
  to { content: none; }
}

#target:before {
  content: 'visible';
  animation: no-content 5s infinite;
}
```

Demo at: https://jsbin.com/nuwagoc/edit?css,output

When content animates to `none` there is no box generated AFAICT.

**Comment by flackr** [comment]
> * What about introducing a new value that's similar to none, except the boxes are skipped during layout/painting/a11y, but not actually destroyed? Then in animations you could have 'display: none' (and 'content: none' on pseudos) behave as that value.

I think this is effectively `content-visibility: hidden`, which developers should also be able to do as part of this change (i.e. we should also make content-visibility animatable as discrete interpolation.

> * Alternatively, since the main use case seems to be animating to/from 'display: none', introducing a new property (or a new value of 'visibility'?) that flips between 'display me' and 'don't display me' and having _that_ be animatable as discrete. (There's been requests for such a property in the past, to avoid messing with the "which box type" aspect of 'display'.)

If we have interpolation from any value and `none` always choose the non-none value (similar to visibility animations) we will have effectively this. 

> Rebuilding layout trees is not fast, I'm surprised you want to enable it during an animation.

The flip of the value happens once (per keyframe pair that changes the value), this is simply enabling developers to coordinate these property changes with the rest of their animation without having to do the same style change / layout tree rebuild they would have done in a JS listener.

It does seem worth avoiding potential cycles in determining what the state should be - I just thought it was interesting that we already had allowed what is essentially a cycle with content animations. Interestingly on my demo every browser handles it differently:
* Chrome applies the `content: none` and never updates again (e.g. the animation is effectively canceled but none is still applied.
* Firefox continues the animation even after the box goes away.
* Safari applies the `content: none` and which removes the pseudo for a frame but then restyles it as the animation override is no longer there restarting the animation. This is the cycle we are trying to avoid by having `none` compute to the underlying style.

**Comment by fantasai** [comment]
> I think this is effectively content-visibility: hidden

No, `content-visibility` only affects the content not the box itself. Right?

**Comment by fantasai** [comment]
> The flip of the value happens once (per keyframe pair that changes the value)

You said in https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1453876094 that only `display: none` gets ignored, an animation flipping back and forth between `inline` and `grid` would therefore still be possible.

**Comment by flackr** [comment]
I agree, it would be a bad idea to animate `display` between inline and grid repeatedly. Typically however, we make properties [not animatable](https://www.w3.org/TR/web-animations-1/#not-animatable) for fundamental spec complexity, rather than efficiency arguments. Animating between non-none display values is well defined in how it should apply and I feel like it would just add more spec complexity to try to carve out a special case for a property we already have and developers already use to flip between those states.

**Comment by fantasai** [comment]
OK, edits are in https://drafts.csswg.org/css-display-4/#display-animation

What do you want to do about `content: none`? Should we replace it with `content: ""` the same way we're replacing `display: none` with `display: revert-layer`?

**Comment by cdoublev** [comment]
> [...] a declared value of `none` in an animation effect or `@keyframes` rule is replaced with `revert-layer`.

Given the above definition, please correct me if the following expectation is wrong:

```html
<style>
  @keyframes myAnimation { to { display: none } }
</style>
<script>
  document.styleSheets[0].cssRules[0].cssRules[0].style.display; // 'revert-layer'
</script>
```

**Comment by flackr** [comment]
> OK, edits are in https://drafts.csswg.org/css-display-4/#display-animation

One slight question - none only interrupts CSS animations, so I think we only need to modify `none` values in animations with an [owning element](https://drafts.csswg.org/css-animations-2/#owning-element).

> What do you want to do about `content: none`? Should we replace it with `content: ""` the same way we're replacing `display: none` with `display: revert-layer`?

Yeah we should probably do the same with content as it has exactly the same sort of problem and currently has 3 completely different behaviors across browsers. Again, only necessary when the animation has an owning element.

**Comment by fantasai** [comment]
> Given the above definition, please correct me if the following expectation is wrong

I think so? @flackr @emilio, is this what you expect?

> One slight question - none only interrupts CSS animations, so I think we only need to modify none values in animations with an [owning element](https://drafts.csswg.org/css-animations-2/#owning-element).

I'm not sure what you're trying to optimize here. Does an animation have any observable effect if it doesn't have an owning element?

> Yeah we should probably do the same with content as it has exactly the same sort of problem and currently has 3 completely different behaviors across browsers.

Filed https://github.com/w3c/csswg-drafts/issues/8623

**Comment by fantasai** [comment]
@flackr OK. I think I'd like to confirm with the other implementers what is the easiest set of restrictions to implement, whether it's easier to be more targetted about it or to be more broad about it. CC @emilio @smfr

**Question:** We decided to replace `display: none` in keyframes with `display: revert-layer` in order to avoid an animation setting a style that stops its own animation. When/where/under what conditions exactly should this replacement take place?

See also follow-up issues on `content` and `content-visibility`:
* https://github.com/w3c/csswg-drafts/issues/8623
* https://github.com/w3c/csswg-drafts/issues/8627

**Comment by hiikezoe** [comment]
Though it's not a strong preference, I personally prefer the alternative approach in https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1462399688 .  The example of CSS animation in https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1318933547 looks awkward to me. Given that to animate the display property web authors need to specify at least a display property value in a keyframe, even if it's the same right? For example, the example can be written;

```
@keyframes slideaway {
  from { display: flex; }
  to { transform: translateY(40px); opacity: 0;}
}

#target {
  display: flex;
}

#target.hide {
  animation: slideaway 200ms;
  display: none;
}
```

Web authors will probably need to care that they need to specify the same display value. Am I missing something?

**Comment by hiikezoe** [comment]
Okay so, web authors normally will write something like;

```
@keyframes slideaway {
  to { transform: translateY(40px); opacity: 0; display: none:}
}

#target.hide {
  animation: slideaway 200ms;
  display: none;
}
```
This doesn't look awkward. :)

**Comment by KevinDoughty** [comment]
Instead of removing animations when set to none, remove animations when the computed value resolves to none. A developer could animate display by setting a transition with a step-end easing on the display property. But also, create a class with an animation on the display property with keyframes from block to block. The animation should have a higher composite order than the transition on the same element so it overrides the transition. Animations would still need to be triggered when the specified value of display is set, instead of the computed value which is kind of arbitrary. But it doesn't have to be additive, and doesn't require adding new keywords like inert https://github.com/w3c/csswg-drafts/issues/8389

**Comment by flackr** [comment]
So when I was experimenting with animating `content` it seemed like Firefox didn't stop animating the pseudo-element even when content was none. This got me wondering whether we could only remove animations if the base computed display style is none. This might I think fix the circularity issues and allow the feature to work the simple way that a developer would expect.

**Comment by flackr** [comment]
> Okay so, web authors normally will write something like;
> 
> ```
> @keyframes slideaway {
>   to { transform: translateY(40px); opacity: 0; display: none:}
> }
> 
> #target.hide {
>   animation: slideaway 200ms;
>   display: none;
> }
> ```
> 
> This doesn't look awkward. :)

@hiikezoe this doesn't work. The implicit `from` keyframe uses the underlying value for display, which is `none` from the `#target.hide` rule. `transition: display 200ms` however would work.

**Comment by hiikezoe** [comment]
> > Okay so, web authors normally will write something like;
> > ```
> > @keyframes slideaway {
> >   to { transform: translateY(40px); opacity: 0; display: none:}
> > }
> > 
> > #target.hide {
> >   animation: slideaway 200ms;
> >   display: none;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This doesn't look awkward. :)
> 
> @hiikezoe this doesn't work. The implicit `from` keyframe uses the underlying value for display, which is `none` from the `#target.hide` rule. `transition: display 200ms` however would work.

Oops, you are right. :/

**Comment by josepharhar** [comment]
> I think I'd like to confirm with the other implementers what is the easiest set of restrictions to implement, whether it's easier to be more targetted about it or to be more broad about it.

> So when I was experimenting with animating content it seemed like Firefox didn't stop animating the pseudo-element even when content was none. This got me wondering whether we could only remove animations if the base computed display style AND the computed display style is none. This might I think fix the circularity issues and allow the feature to work the simple way that a developer would expect.

I have implemented both display:none->display:revert and flackr's idea here of not canceling the animation when the animation sets the style to display:none. Unless there are edge cases I'm not thinking of, both were similar in difficulty to implement. I don't have a preference yet of which we should go with, but I defer to flackr's judgement.

To implement display:none->display:revert, I modified the style resolution code in parts where it knows that the style is being applied for an animation to change display:none to display:revert. It required additional changes in another spot to handle this case with variables:
```css
@keyframes hello {
  0% { --display-value: none; }
  100% { --display-value: none; }
}
#target {
  display: var(--display-value, block);
}
#target.animation {
  animation-name: hello;
  animation-duration: 1s;
}
```

To implement not canceling the animation when an animation sets display:none, I used similar spots in the style resolution code but set a flag on the element when the animation sets display:none, so that the element knows that it shouldn't cancel the animation right after it recalcs its own style.

## use cases

Not canceling the animation provides the ability to animate to display:none and back with keyframes during an animation and actually make the element disappear, but I don't know if that's a use case that we need to support. Here is an example of what that would look like:
```css
@keyframes {
  0% { display: block; }
  20% { display: none; }
  80% { display: none; }
  100% { display: block; }
}
```

> Display is interpolated like visibility, i.e. it will prefer the non-none value.

Due to this `visibility`-like behavior, **none of the following use cases will be affected by the display:none->display:revert or don't cancel the animation behavior** because there has to be a full interpolation step to and from display:none. All of these examples will be display:block for the entire duration of the animation regardless of which implementation option we choose.

```css
@keyframes {
  0% { display: block; }
  100% { display: none; }
}
```
```css
@keyframes {
  0% { display: none; }
  100% { display: block; }
}
```
```css
#target {
  display: block;
  transition: display 1s;
}
#target.animated { display: none; }
```
```css
#target {
  display: none;
  transition: display 1s;
}
#target.animated { display: block; }
```

**Comment by flackr** [comment]
Thanks @josepharhar for the implementation experience!

I believe that changing the cancellation rules for css animations to be only when the base computed style *and* the animated display style both produce `none` is much simpler to understand and explain, so I'd like to propose going with this since the implementation experience suggests that this works (and also that it has fewer special cases).

This is also consistent with the way that animations of `content` to none on pseudo-elements behave in Firefox today (animation keeps running), and conceptually similar to animating visibility.

**Comment by josepharhar** [comment]
> To implement not canceling the animation when an animation sets display:none, I used similar spots in the style resolution code but set a flag on the element when the animation sets display:none, so that the element knows that it shouldn't cancel the animation right after it recalcs its own style.

> the implementation experience suggests that this works (and also that it has fewer special cases).

Yeah it turns out that not canceling the animation is much easier to implement, I actually don't need to mess with the style resolution code at all that I mentioned earlier like I did in order to implement the display:revert behavior

**Comment by chrishtr** [comment]
Agenda+ to resolve on the behavior Rob suggested [here](https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1503849651).

**Comment by KevinDoughty** [comment]
There is no distinction between computed and animated values, is there? Aren't they one and the same? I asked for this distinction a decade ago in public-fx, because the pattern I call relative animation doesn't need current animated values for interruption. It only ever needs discrete values. Also, the underlying value is not necessarily discrete, which also confused me ten years ago.

Is or is not the comment I made almost eight hours before flackr the same solution?

This is what I wrote: https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1486246609
> Instead of removing animations when set to none, remove animations when the computed value resolves to none.

This is what flackr wrote less than eight hours later: https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1486895943
> This got me wondering whether we could only remove animations if the base computed display style AND the computed display style is none.

And then yesterday: https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1503849651
> only when the base computed style _and_ the animated display style both produce `none`

Can someone please explain the difference? If not, please make the distinction between discrete and animated computed values in specification. Is there a new definition in css-cascade-4, css-cascade-5, or css-cascade-6? They're hard to read since they are in some .bs format (no pun intended).

Another question, will something like this be possible? https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1486246609
> animate display by setting a transition with a step-end easing on the display property. But also, create a class with an animation on the display property with keyframes from block to block.

It would make a lot of sense for front-end developers:
```
@keyframes displaying {
    0% { display: block; }
    100% { display: block; }
}
#target {
    transition: display 1s step-end, opacity 1s linear;
    animation: displaying 1s linear;
}
.in {
    display: block;
    opacity: 1;
}
.out {
    display: none;
    opacity: 0;
}
```
Would this work? Can we please make it work? If not, why?

**Comment by flackr** [comment]
> There is no distinction between computed and animated values, is there? Aren't they one and the same? I asked for this distinction a decade ago in public-fx, because the pattern I call relative animation doesn't need current animated values for interruption. It only ever needs discrete values. Also, the underlying value is not necessarily discrete, which also confused me ten years ago.
> 
> Is or is not the comment I made almost eight hours before flackr the same solution?
> 
> This is what I wrote: [#6429 (comment)](https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1486246609)
> 
> > Instead of removing animations when set to none, remove animations when the computed value resolves to none.
> 
> This is what flackr wrote less than eight hours later: [#6429 (comment)](https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1486895943)
> 
> > This got me wondering whether we could only remove animations if the base computed display style AND the computed display style is none.


> 
> And then yesterday: [#6429 (comment)](https://github.com/w3c/csswg-drafts/issues/6429#issuecomment-1503849651)
> 
> > only when the base computed style _and_ the animated display style both produce `none`
> 
> Can someone please explain the difference? If not, please make the distinction between discrete and animated computed values in specification. Is there a new definition in css-cascade-4, css-cascade-5, or css-cascade-6? They're hard to read since they are in some .bs format (no pun intended).

The **base** word is the key difference here. This is referring to the style before the animations and transitions cascade. The computed value will resolve to `none` if we allow the animation to include `display: none` in its keyframes which then leads to the problematic circularity.

However, if the value before the animations part of the cascade is also `none` then when the computed value becomes `none` canceling the animation will not restart the animation.

I've taken a pass at writing the spec updates for this in https://github.com/w3c/csswg-drafts/pull/8713

**Comment by KevinDoughty** [comment]
At the risk of sealioning, I just want to say this will probably not work as is with my (experimental and unapproved by the W3C) desired changes for discrete state:
https://phabricator.services.mozilla.com/D156634#5139009

These are flackr's changesets and diffs for my or anyone else's future reference:
https://chromium-review.googlesource.com/c/chromium/src/+/4356155
https://chromium.googlesource.com/chromium/src/+/40601f60e9d413ca1fad4891041c0f43925a7f7e%5E%21/
https://chromium-review.googlesource.com/c/chromium/src/+/4363078
https://chromium.googlesource.com/chromium/src/+/1cc90dd5bb6ba48e9b359d3e93e8223e7eae98de%5E%21/

**Comment by flackr** [comment]
> At the risk of sealioning, I just want to say this will probably not work as is with my (experimental and unapproved by the W3C) desired changes for discrete state: https://phabricator.services.mozilla.com/D156634#5139009

Can you be more specific? In particular,
* Can you confirm "this" the proposal for allowing animations to continue as long as their non-animated display is not none?
* What specifically makes it not work with the discrete state changes? I didn't quite follow from reading the linked proposal why it doesn't work?

**Comment by KevinDoughty** [comment]
> Can you be more specific?

I've been incapacitated, again, but will try to make a coherent proposal to sustyweb.

**Comment by emilio** [comment]
Does this still need spec edits?
